packages/sdk38/src/lcdclient.ts

```diff
@@ -0,0 +1,350 @@
+  NodeInfoResponse,
+  PostTxsResponse,
+  SearchTxsResponse,
+  TxsResponse,
+} from "./restclient";
+import { CosmosSdkTx, StdTx } from "./types";
+
+/** Unfortunately, Cosmos SDK encodes empty arrays as null */
+export type LcdApiArray<T> = ReadonlyArray<T> | null;
+
+export function normalizeArray<T>(backend: LcdApiArray<T>): ReadonlyArray<T> {
+  return backend || [];
+}
+
+type LcdModule = Record<string, () => any>;
+
+type LcdModuleSetup<M> = (base: LcdClient) => M;
+
+export interface LcdClientBaseOptions {
+  readonly apiUrl: string;
+  readonly broadcastMode?: BroadcastMode;
+}
+
+// We want to get message data from 500 errors
+// https://stackoverflow.com/questions/56577124/how-to-handle-500-error-message-with-axios
+// this should be chained to catch one error and throw a more informative one
+function parseAxiosError(err: AxiosError): never {
+  // use the error message sent from server, not default 500 msg
+  if (err.response?.data) {
+    let errorText: string;
+    const data = err.response.data;
+    // expect { error: string }, but otherwise dump
+    if (data.error && typeof data.error === "string") {
+      errorText = data.error;
+    } else if (typeof data === "string") {
+      errorText = data;
+    } else {
+      errorText = JSON.stringify(data);
+    }
+    throw new Error(`${errorText} (HTTP ${err.response.status})`);
+  } else {
+    throw err;
+  }
+}
+
+/**
+ * A client to the LCD's (light client daemon) API.
+ * This light client connects to Tendermint (i.e. the chain), encodes/decodes Amino data for us and provides a convenient JSON interface.
+ *
+ * This _JSON over HTTP_ API is sometimes referred to as "REST" or "RPC", which are both misleading terms
+ * for the same thing.
+ *
+ * Please note that the client to the LCD can not verify light client proofs. When using this,
+ * you need to trust the API provider as well as the network connection between client and API.
+ *
+ * @see https://cosmos.network/rpc
+ */
+export class LcdClient {
+  /** Constructs an LCD client with 0 modules */
+  public static withModules(options: LcdClientBaseOptions): LcdClient;
+
+  /** Constructs an LCD client with 1 module */
+  public static withModules<A extends LcdModule>(
+    options: LcdClientBaseOptions,
+    setupModuleA: LcdModuleSetup<A>,
+  ): LcdClient & A;
+
+  /** Constructs an LCD client with 2 modules */
+  public static withModules<A extends LcdModule, B extends LcdModule>(
+    options: LcdClientBaseOptions,
+    setupModuleA: LcdModuleSetup<A>,
+    setupModuleB: LcdModuleSetup<B>,
+  ): LcdClient & A & B;
+
+  /** Constructs an LCD client with 3 modules */
+  public static withModules<A extends LcdModule, B extends LcdModule, C extends LcdModule>(
+    options: LcdClientBaseOptions,
+    setupModuleA: LcdModuleSetup<A>,
+    setupModuleB: LcdModuleSetup<B>,
+    setupModuleC: LcdModuleSetup<C>,
+  ): LcdClient & A & B & C;
+
+  /** Constructs an LCD client with 4 modules */
+  public static withModules<
```

Comment:
It wouldn't be easier to just use `withModules(options: LcdModule, modules: ReadonlyArray<LcdModule>)`?

You need to keep the specific type info?

Reply:
The difference is the type safety of the result. This way, the client from `const client = LcdClient.withModules({ apiUrl: wasmd.endpoint }, registerWasmModule, setupSupplyModule);` is a `LcdClient & WasmModule & SupplyModule`, providing all methods for all of them.

Reply:
Ahhhh... that is your trick to get the types proper. I was wondering what was going on.